In [41]:
# Initialize Otter
import otter
grader = otter.Notebook("S03_Structuring_Code_and_Data.ipynb")

# Session 3: Structuring Code and Data

Functions organize instructions into named, reusable blocks. Modules and packages organize code into files and directories. Data structures organize related values.

## Learning objectives

After completing this session, you can:
- describe a function using Input–Processing–Output;
- define parameters, pass arguments, and use a returned value;
- distinguish returning a value from printing text;
- create Python modules and import them into a notebook;
- create regular packages and a subpackage using `__init__.py`;
- use lists, tuples, sets, and dictionaries for simple tasks.

Run the notebook from top to bottom. After each exercise, execute its `grader.check(...)` cell for feedback. Blue checkpoints are for discussion; yellow exercises have automated tests.


In [ ]:
from pathlib import Path
import sys
print("Create your exercise files and folders in:", Path.cwd())


## 3.1 Functions

A function definition names a reusable block of instructions. Parameters receive the argument values supplied in a call. `return` sends a value back to the caller; `print()` displays text.

```python
def delivery_fee(distance):
    return 2.0 + distance * 0.5

fee = delivery_fee(4)
print(f"CHF {fee:.2f}")  # CHF 4.00
```

A function can have no parameters. Without an explicit return value, Python returns `None`.


<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #3b82f6;background: rgba(59, 130, 246, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Checkpoint 3.1: A function as IPO</h3>
<p>For the example above, identify the parameter, argument, processing, and returned value. Which line displays the result? Why is displaying a result different from returning it?</p><p><em>Write your answers below and discuss them with a neighbour. This checkpoint is not automatically graded.</em></p>
</div>


<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.1.1: Calculate an order total</h3>
<p><strong>Points:</strong> 3</p>
<p>Create a function named <code>calculate_total</code> with two parameters:</p>
<ul>
<li><code>price</code>: the unit price as a float.</li>
<li><code>quantity</code>: the number of items as a non-negative integer.</li>
</ul>
<p>The function should multiply the unit price by the quantity and return the result. Do not call <code>input()</code> or <code>print()</code> inside the function.</p>
<p>Example: <code>calculate_total(12.50, 2)</code> returns <code>25.0</code>.</p>
</div>

In [8]:
# Write your code below this line
def calculate_fee(amount, fee_percentage):
    return amount * fee_percentage / 100

calculated = calculate_fee(12.50, 2)
print(f"CHF {calculated:.2f}")


CHF 0.25


In [ ]:
grader.check("3.1.1")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.1.2: Call a function with different arguments</h3>
<p><strong>Points:</strong> 3</p><p>Use your function from Exercise 3.1.1 twice. Assign the total for <code>2</code> items at CHF <code>12.50</code> to <code>first_total</code>, and the total for <code>3</code> items at CHF <code>14.00</code> to <code>second_total</code>. Assign their sum to <code>combined_total</code>. Use the returned values rather than typing the calculated answers.</p>
</div>


In [13]:
# Write your code below this line
def calculate_fee(amount, fee_percentage):
    return amount + fee_percentage

first_total = calculate_fee(12.50, 12.50)

def calculate_fee2(amount, fee_percentage):
    return amount + fee_percentage

second_total = calculate_fee2(25.00, 25.00)
combined_total = first_total + second_total

print(f"CHF {combined_total:.2f}")

CHF 75.00


In [ ]:
grader.check("3.1.2")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.1.3: Calculate VAT using parameters</h3>
<p><strong>Points:</strong> 3</p><p>Define <code>calculate_vat(price_inc_vat, vat_rate)</code>. Return the vat amount without printing. Both arguments are floats. Use the supplied rate rather than fixing one rate inside the function.</p><p>Call it with <code>25.0</code> and <code>0.026</code>, assigning the returned value to <code>vat_amount</code>.</p>

<p><strong>Hint:</strong> The price_inc_vat already includes VAT. At a VAT rate of 2.6%, it represents 102.6% of the net amount. Divide by <code>1 + vat_rate</code> to find the net amount, then subtract it from the original amount to obtain the VAT amount.</p>
</div>


In [32]:
# Write your code below this line
def calculate_vat(amount, vat_percentage):
    return amount * vat_percentage
vat_amount = calculate_vat(25, 0.026)

print(vat_amount)


0.65


In [42]:
grader.check("3.1.3")

3.1.3 results:
    3.1.3 - 1 message: Review the assertion message and the exercise requirements.

    3.1.3 - 1 result:
        ❌ Test case failed
        Error at line 19 in test 3.1.3:
              assert list(parameters) == ['price_inc_vat', 'vat_rate'], 'The parameters of calculate_vat must be named price_inc_vat and vat_rate.'
        AssertionError: The parameters of calculate_vat must be named price_inc_vat and vat_rate.

    3.1.3 - 2 message: Review the assertion message and the exercise requirements.

    3.1.3 - 2 result:
        ❌ Test case failed
        Error at line 34 in test 3.1.3:
              assert vat_amount == 25.0 - 25.0 / (1 + 0.026), 'Check the value of vat_amount.'
        AssertionError: Check the value of vat_amount.

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.1.4: Return text and display it</h3>
<p><strong>Points:</strong> 3</p><p>Complete <code>format_currency(amount)</code> so it returns a string such as <code>"CHF 25.00"</code>. Use exactly two decimal places. Then complete <code>show_total(amount)</code>: call <code>format_currency</code> and print its returned text once. <code>show_total</code> should have no explicit return value.</p>
</div>


In [31]:
def format_currency(amount):
    return f"CHF {amount:.2f}"


def show_total(amount):
    print (format_currency(amount))


show_total(25)


CHF 25.00


In [43]:
grader.check("3.1.4")

3.1.4 results: All test cases passed!
3.1.4 - 1 message: The result is correct.
3.1.4 - 2 message: The result is correct.

## 3.2 Modules

A module stores related definitions in a Python file. Use `import module_name` and access its definitions with `module_name.name`. Alternatively, `from module_name import function_name` imports a function directly.

For the next exercises, create `pricing.py`, `formatting.py`, and `receipt.py` **beside this notebook**. Keep calculations in `pricing.py`, currency formatting in `formatting.py`, and receipt composition in `receipt.py`. These files must work without variables defined in this notebook. Importing them must not ask for input or print output.

Exercises 3.2.1–3.2.4 build on each other. Keep earlier definitions when adding new ones.


<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #3b82f6;background: rgba(59, 130, 246, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Checkpoint 3.2: Separate responsibilities</h3>
<p>Why do calculations belong in <code>pricing.py</code> and text formatting in <code>formatting.py</code>? Which file would you change to display EUR instead of CHF? Explain the difference between <code>import pricing</code> and <code>from pricing import calculate_total</code>.</p><p><em>Write your answers below and discuss them with a neighbour. This checkpoint is not automatically graded.</em></p>
</div>


<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.2.1: Create and import your first module</h3>
<p><strong>Points:</strong> 3</p><ol><li>Create <code>pricing.py</code> beside the notebook.</li><li>Copy your <code>calculate_total(price, quantity)</code> function into it. It must work independently of the notebook.</li><li>In the notebook, import the module with <code>import pricing</code>.</li><li>Use <code>pricing.calculate_total</code> for <code>4</code> items at CHF <code>4.50</code> and assign its result to <code>module_total</code>.</li></ol>
</div>


In [ ]:
# Create or update pricing.py in your editor and save it.
sys.modules.pop("pricing", None) # This ensures that the module is reloaded and any changes made to pricing.py are reflected in the current session.

# Write your code below this line
# Because the module has been removed, you need to re-import it to access its new contents.

...


In [ ]:
grader.check("3.2.1")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.2.2: Add constants and a function to your module</h3>
<p><strong>Points:</strong> 2</p><p>Update <code>pricing.py</code>. Keep <code>calculate_total</code>, add <code>calculate_vat(price_inc_vat, vat_rate)</code>, and define <code>VAT_NORMAL = 0.081</code>, <code>VAT_REDUCED = 0.026</code>, and <code>VAT_NONE = 0.0</code>.</p><p>Save the file and run the cell. Use <code>pricing.VAT_REDUCED</code> to calculate VAT on CHF 25.00 and assign it to <code>module_vat</code>.</p>
</div>


In [ ]:
# Create or update pricing.py in your editor and save it.
import sys
sys.modules.pop("pricing", None) # This ensures that the module is reloaded and any changes made to pricing.py are reflected in the current session.

# Write your code below this line
# Because the module has been removed, you need to re-import it to access its new contents.

...


In [ ]:
grader.check("3.2.2")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.2.3: Import a function directly</h3>
<p><strong>Points:</strong> 3</p><p>Create <code>formatting.py</code> with <code>format_currency(amount)</code>. It returns text such as <code>"CHF 7.50"</code> without printing.</p><p>In the notebook, use <code>from formatting import format_currency</code>. This replaces the earlier notebook definition of that name. Call the imported function for <code>7.5</code> and assign its result to <code>formatted_price</code>.</p>
</div>


In [ ]:
import sys
# Create or update pricing.py in your editor and save it.
sys.modules.pop("formatting", None) # This ensures that the module is reloaded and any changes made to pricing.py are reflected in the current session.

# Write your code below this line
# Because the module has been removed, you need to re-import it to access its new contents.
...


In [ ]:
grader.check("3.2.3")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.2.4: Reuse two modules</h3>
<p><strong>Points:</strong> 3</p><p>Create <code>receipt.py</code>. Import <code>pricing</code> and <code>formatting</code>. Define <code>format_receipt(price, quantity)</code> to call their existing functions and return the formatted total. Reuse the functions rather than repeating their calculations or formatting code.</p><p>In the notebook, import <code>receipt</code>. Call its function with <code>12.50</code> and <code>2</code>, storing the result in <code>receipt_text</code>. Expected result: <code>"CHF 25.00"</code>.</p>
</div>


In [ ]:
# Write your code below this line
...


In [ ]:
grader.check("3.2.4")

## 3.3 Packages

A package groups related modules and subpackages. For these exercises, create **packages** by adding an `__init__.py` file to each package directory. Keep these files empty or containing only a comment. Python executes initialization code when a package is first imported.

Create the following structure beside the notebook as you work through this section:

```text
./
├── S03_Structuring_Code_and_Data.ipynb
├── logic/
│   ├── __init__.py
│   ├── pricing.py
│   ├── checkout.py
│   └── delivery/
│       ├── __init__.py
│       └── fees.py
└── user_interface/
    ├── __init__.py
    └── formatting.py
```

Keep the earlier top-level files. `pricing` and `logic.pricing` are different module names. In a real application, we would move the module into the package and not keep a copy.


<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.3.1: Create a package and import its module</h3>
<p><strong>Points:</strong> 3</p><p>Create <code>logic/</code>, add <code>logic/__init__.py</code>, and copy your complete <code>pricing.py</code> into <code>logic/pricing.py</code>. The copied module must contain its own definitions, so it works even without the top-level <code>pricing.py</code>.</p><p>Use <code>import logic.pricing</code> in the notebook. Call <code>logic.pricing.calculate_total(7.25, 2)</code> and assign the result to <code>package_total</code>.</p>
</div>


In [37]:
# Write your code below this line
import sys
sys.path.insert(0, "..")
import logic.pricing

logic.pricing.calculate_total(7.25, 2)




14.5

In [38]:
grader.check("3.3.1")

NameError: name 'grader' is not defined

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.3.2: Create a package for presentation code</h3>
<p><strong>Points:</strong> 3</p><p>Create the regular package <code>user_interface</code> with <code>__init__.py</code>. Add <code>user_interface/formatting.py</code> containing your <code>format_currency(amount)</code> definition.</p><p>Use <code>import user_interface.formatting</code> in the notebook. Call <code>user_interface.formatting.format_currency(14.5)</code> and store the result in <code>package_text</code>.</p>
</div>


In [ ]:
# Write your code below this line
...


In [ ]:
grader.check("3.3.2")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.3.3: Create a subpackage</h3>
<p><strong>Points:</strong> 3</p><p>Inside <code>logic</code>, create a directory <code>delivery</code> with its own <code>__init__.py</code>. In <code>logic/delivery/fees.py</code>, define <code>BASE_FEE = 2.0</code>, <code>PER_KM = 0.5</code>, and <code>delivery_cost(distance)</code>. Return the base fee plus the distance multiplied by the per-kilometre fee. Use the constants.</p><p>In the notebook, use <code>import logic.delivery.fees</code>. Calculate the fee for <code>4</code> kilometres and assign it to <code>delivery_amount</code>. Expected result: <code>4.0</code>.</p>
</div>


In [ ]:
# Write your code below this line
...


In [ ]:
grader.check("3.3.3")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.3.4: Connect modules inside a package</h3>
<p><strong>Points:</strong> 3</p><p>Create <code>logic/checkout.py</code>. Import the <code>pricing</code> module from <code>logic</code> and the <code>fees</code> module from <code>logic.delivery</code>. Define <code>calculate_checkout(price, quantity, distance)</code> to call both functions and return the sum of the item total and delivery fee. Do not duplicate their formulas. No VAT is added in this exercise.</p><p>In the notebook, import <code>checkout</code> from <code>logic</code>. Calculate two items at CHF 12.50 with delivery over 4 km. Assign the result to <code>checkout_total</code>. Use <code>format_currency</code> from Exercise 3.3.2 to assign the formatted total to <code>checkout_text</code>. Expected text: <code>"CHF 29.00"</code>.</p>
</div>


In [ ]:
# Write your code below this line
...


In [ ]:
grader.check("3.3.4")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #3b82f6;background: rgba(59, 130, 246, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Checkpoint 3.3: Files and folders</h3>
<p>Identify one function, one module, one package, and one subpackage in your project. Why does <code>logic/delivery</code> have its own <code>__init__.py</code>? Which imports distinguish <code>pricing.py</code> from <code>logic/pricing.py</code>?</p><p><em>Write your answers below and discuss them with a neighbour. This checkpoint is not automatically graded.</em></p>
</div>


## 3.4 Data structures

| Structure | Main properties | Example |
|---|---|---|
| `list` | Ordered, mutable, permits duplicates | A changing basket |
| `tuple` | Ordered, immutable, permits duplicates | Coordinates |
| `set` | Unordered, mutable, unique values | Product tags |
| `dict` | Key–value pairs, mutable, unique keys, preserves insertion order | Prices by product name |

**Ordered** means items have defined positions; it does not mean sorted by value. A tuple's items cannot be added, removed, or replaced. A dictionary can contain the same value under different keys.

Small syntax examples:
```python
colours = ["red", "blue"]
colours.append("green")
first_colour = colours[0]      # indexes start at zero
point = (4, 8)
x = point[0]
tags = {"new", "local"}
tags.add("new")                # duplicates do not add another entry
is_local = "local" in tags
prices = {"tea": 3.0}
prices["tea"] = 3.5
tea_price = prices["tea"]
```

The four exercises below are independent and do not need the files from earlier sections.


<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #3b82f6;background: rgba(59, 130, 246, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Checkpoint 3.4: Choose a suitable collection</h3>
<p>Choose a data structure for each case and justify your choice: a changing shopping basket, a fixed coordinate pair, unique product tags, and prices found by product name. Explain why a set is unsuitable when you must count repeated orders of the same pizza.</p><p><em>Write your answers below and discuss them with a neighbour. This checkpoint is not automatically graded.</em></p>
</div>


<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.4.1: Create a pizza menu</h3>
<p><strong>Points:</strong> 1</p>

<p>
Create a list named <code>menu</code> containing the two strings
<code>"margherita"</code> and <code>"salami"</code>, in that order.
</p>

</div>

In [51]:
# Write your code below this line
menu = list(["margherita", "Lasami", "REGEN"])

TypeError: 'list' object is not callable

In [50]:
grader.check("3.4.1")

3.4.1 results:
    3.4.1 - 1 message: Create menu as a list containing the required pizzas in the correct order.

    3.4.1 - 1 result:
        ❌ Test case failed
        Error at line 13 in test 3.4.1:
              assert menu[:2] == ['margherita', 'salami'], "menu should contain 'margherita' and 'salami' in that order."
        AssertionError: menu should contain 'margherita' and 'salami' in that order.

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.4.2: Add a pizza to the menu</h3>
<p><strong>Points:</strong> 1</p>

<p>
Use <code>append()</code> to add <code>"funghi"</code> to the existing
<code>menu</code> list.
</p>

<p>
Afterwards, <code>menu</code> should contain
<code>"margherita"</code>, <code>"salami"</code>, and <code>"funghi"</code>,
in that order.
</p>
</div>

In [ ]:
# Write your code below this line
...


In [ ]:
grader.check("3.4.2")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.4.3: Store a fixed order item</h3>
<p><strong>Points:</strong> 1</p>

<p>
Create a tuple named <code>ordered_pizza</code> containing the pizza name
<code>"margherita"</code> and its agreed price <code>12.50</code>, in that order.
</p>

<p>
Use a tuple because the pizza and its price represent a fixed snapshot of the
order. The menu price might change later, but the existing order must remain
unchanged.
</p>

<p>
The expected value is:
<code>("margherita", 12.50)</code>
</p>
</div>

In [ ]:
# Write your code below this line
ordered_pizza = ...


In [ ]:
grader.check("3.4.3")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 20px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.4.4: Work with unique pizza tags</h3>
<p><strong>Points:</strong> 2</p>

<p>
Create a function named <code>pizza_tags()</code>.
</p>

<p>
Inside the function, create a set containing
<code>"vegetarian"</code> and <code>"gluten-free"</code>.
Print the set, use <code>.add()</code> to add <code>"vegetarian"</code>
again, and print the set a second time.
</p>

<p>
Discuss with your neighbour, why the set did not change after the <code>add()</code>.
</p>

<p>
Finally, return the set from the function.
</p>

<p>
Call <code>pizza_tags()</code> and assign the returned set to
<code>available_tags</code>. Use <code>in</code> to determine whether
<code>"vegetarian"</code> is present and assign the result to
<code>is_vegetarian</code>.
</p>
</div>

In [ ]:
# Write your code below this line
...

available_tags = ...
is_vegetarian = ...



In [ ]:
grader.check("3.4.4")

<div style="box-sizing:border-box;max-width:100%;width:calc(100% - 10px);border-left:6px solid #f59e0b;background: rgba(240, 160, 0, 0.08);padding:16px 20px;margin:18px 10px 18px 0;border-radius:6px;">
<h3 style="margin:0 0 12px;">Exercise 3.4.5: Look up menu prices</h3>

<p><strong>Points:</strong> 2</p>

<p>
Create a dictionary named <code>menu_prices</code> with the following
pizza prices:
</p>

<ul>
  <li><code>"margherita"</code>: <code>12.50</code></li>
  <li><code>"salami"</code>: <code>14.50</code></li>
  <li><code>"funghi"</code>: <code>13.00</code></li>
</ul>

<p>
Use dictionary lookups to assign the price of
<code>"margherita"</code> to <code>margherita_price</code> and the price of
<code>"salami"</code> to <code>salami_price</code>.
</p>
</div>

In [ ]:
# Write your code below this line
...
margherita_price = ...
salami_price = ...



In [ ]:
grader.check("3.4.5")

## References

Course slides: *Session 3: Structuring Code and Data*.

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [ ]:
grader.check_all()

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit. **Please save before exporting!**

In [ ]:
# Save your notebook first, then run this cell to export your submission.
grader.export(pdf=False, run_tests=True)